src/facebook/facebook-images/facebookimagesyncadaptor.cpp

```diff
@@ -308,29 +308,48 @@ void FacebookImageSyncAdaptor::imagesFinishedHandler()
         QString createdTimeStr = imageObject.value(QLatin1String("created_time")).toString();
         QString updatedTimeStr = imageObject.value(QLatin1String("updated_time")).toString();
         QString photoName = imageObject.value(QLatin1String("name")).toString();
+        int imageWidth = 0;
+        int imageHeight = 0;
 
-        // Find the correct thumbnail size. The fallback will be the "picture" which usually
-        // is too small so this is sort of best guess what sizes FB might returns. We can't
-        // also hardcode the exact sizes here, because we can't be sure that certains sizes
-        // will stay for ever.
-        // TODO: we can use https://graph.facebook.com/v2.2/object_id/picture?type=large
+        // Find optimal thumbnail and image source urls based on dimensions.
+        QList<ImageSource> imageSources;
         QJsonArray images = imageObject.value(QLatin1String("images")).toArray();
         foreach (const QJsonValue &imageValue, images) {
             QJsonObject image = imageValue.toObject();
-            int width = static_cast<int>(image.value(QLatin1String("width")).toDouble());
-            int height= static_cast<int>(image.value(QLatin1String("height")).toDouble());
-            if (160 <= width && width <= 350 &&
-                160 <= height && height <= 350) {
-                thumbnailUrl = image.value(QLatin1String("source")).toString();
-                break;
+            imageSources << ImageSource(static_cast<int>(image.value(QLatin1String("width")).toDouble()),
+                                        static_cast<int>(image.value(QLatin1String("height")).toDouble()),
+                                        image.value(QLatin1String("source")).toString());
+        }
+
+        int optimalThumbnailWidth = 360; // if we could access QScreen: QGuiApplication::primaryScreen()->size().width() / 3;
+        int optimalImageWidth = 1536;    // if we could access QScreen: QGuiApplication::primaryScreen()->size().width();
+        bool foundOptimalThumbnail = false, foundOptimalImage = false;
+        std::sort(imageSources.begin(), imageSources.end());
+        Q_FOREACH (const ImageSource &img, imageSources) {
+            if (!foundOptimalThumbnail && qMin(img.width, img.height) >= optimalThumbnailWidth) {
```

Comment:
I'm not 100% sure about the qMin() here, but if we don't do it then we could end up downloading lower-resolution images than would be ideal in the case where the image is landscape oriented.  Note that with this check, we pretty much hit the "fall back to the largest" image every single time, due to the resolutions which FB supports.